File: client-api/src/main/java/org/ostelco/topup/api/resources/AnalyticsResource.java
package org.ostelco.topup.api.resources;

import io.dropwizard.auth.Auth;
import io.vavr.control.Option;
import org.ostelco.topup.api.auth.AccessTokenPrincipal;
import org.ostelco.topup.api.core.Error;
import org.ostelco.topup.api.db.SubscriberDAO;

import javax.validation.constraints.NotNull;
import javax.ws.rs.Consumes;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.core.Response;

/**
 * Analytics API.
 *
 */
@Path("/analytics")
public class AnalyticsResource extends ResourceHelpers {

    private final SubscriberDAO dao;

    public AnalyticsResource(SubscriberDAO dao) {
        this.dao = dao;
    }

    @POST
    @Consumes({"application/json"})
    public Response report(@Auth AccessTokenPrincipal token,
            @NotNull final String event) {
        if (token == null) {
            return Response.status(Response.Status.UNAUTHORIZED)
                .build();
        }

        Option<Error> error = dao.reportAnalytics(token.getName(), event);

        return error.isEmpty()
            ? Response.status(Response.Status.CREATED)
                 .build()
            : Response.status(Response.Status.NOT_FOUND)
                 .entity(asJson(error.get()))
                 .build();
    }
}

File: client-api/src/main/kotlin/org/ostelco/prime/client/api/resources/AnalyticsResource.kt
package org.ostelco.prime.client.api.resources

import io.dropwizard.auth.Auth
import org.ostelco.prime.auth.AccessTokenPrincipal
import org.ostelco.prime.client.api.store.SubscriberDAO
import org.ostelco.prime.jsonmapper.asJson
import org.ostelco.prime.model.Identity
import javax.validation.constraints.NotNull
import javax.ws.rs.Consumes
import javax.ws.rs.POST
import javax.ws.rs.Path
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

/**
 * Analytics API.
 *
 */
@Path("/analytics")
class AnalyticsResource(private val dao: SubscriberDAO) {

    @POST
    @Consumes(MediaType.APPLICATION_JSON)
    fun report(@Auth token: AccessTokenPrincipal?,
               @NotNull event: String): Response {
        if (token == null) {
            return Response.status(Response.Status.UNAUTHORIZED)
                    .build()
        }

        return dao.reportAnalytics(
                identity = Identity(id = token.name, type = "EMAIL", provider = token.provider),
                events = event)
                .fold(
                        { apiError -> Response.status(apiError.status).entity(asJson(apiError)) },
                        { Response.status(Response.Status.CREATED) })
                .build()
    }
}
